### PR TITLE
Rescue hash function

### PR DESCRIPTION
**Pull request by imeckler**
This PR implements the rescue hash function. See [page 22 here](https://eprint.iacr.org/2019/426.pdf) for a spec. I chose the parameters of a state size of 3 with a block size of 2 to be maximally efficient for the Merkle tree. From these parameters, I derived a round size of 11 by checking against all the cryptanalysis described in that paper.

Checking safety against the grobner basis attack with this round size and state size can be checked using the grobner attack complexity given on page 34 plus the estimated degree of regularity given on page 44.

### CODE REVIEW

**Comment by imeckler** [comment]
I'm closing for now because I realized I implemented the MDS matrix sampling incorrectly.

File: src/lib/rescue/rescue.ml [hunk 1]
```diff
@@ -0,0 +1,68 @@
+open Core_kernel
+
+let rounds = 11
```
[the remark this answers is not in the document]
Comment (imeckler) [reply src/lib/rescue/rescue.ml]: Yeah sure. I believe they arrive at this value for the number of rounds in the following way. As mentioned on page 34, the cost of performing the Grobner basis attack is estimated as

( (n + d) choose d ) ^ omega
where 

- omega is some number which is known to be >= 2
- n = 1 + m*N is the number of variables in the system of equations on page 3
- d is a quantity which they estimate as ((alpha - 1)(m*N + 1) + 1) / 2
- m is the state size, which we can choose
- N is the number of rounds which we can choose

They say the attack costs at least 2^(4*m*N) which you can see holds in the limit by taking `alpha = 3` and `omega = 3` in the above formula and doing [Stirling's approximation](https://en.wikipedia.org/wiki/Stirling%27s_approximation). Then taking `N = 2 k / 4` gives an attack complexity of `2^(2 m k)`, and then they give themselves a margin of error and say that this corresponds to a security level of `m k`.

In our case, `alpha = 11`, and I took `m = 3` which is optimal for binary Merkle trees. Evaluating the above formula with these values and `N = 11` and `omega = 2` yields an attack complexity of a little over 2^257, which if we take the same factor of 2 security margin gives us a security level of 257/2 ~= 128.

[the remark this answers is not in the document]
Comment (imeckler) [reply src/lib/rescue/rescue.ml]: Here's a paper which simplifies stirling's approximation for binomials if you want to double check http://page.mi.fu-berlin.de/shagnik/notes/binomials.pdf

File: src/lib/rescue/params.sage [hunk 2]
```diff
@@ -0,0 +1,53 @@
...
+
+def mds():
+    name = prefix + 'MDS'
+    for attempt in xrange(100):
+        x_values = [random_value(name + 'x', attempt * m + i)
+                    for i in xrange(m)]
+        y_values = [random_value(name + 'y', attempt * m + i)
+                    for i in xrange(m)]
+# Make sure the values are distinct.
+        assert len(set(x_values + y_values)) == 2 * m, \
```
[the remark this answers is not in the document]
Comment (imeckler) [reply src/lib/rescue/params.sage]: There being a duplicate is the same thing as a SHA256 collision :)

File: src/lib/rescue/rescue.ml [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+        add_block ~state block ; perm state )
+
+  let sbox0, sbox1 = (alphath_root, to_the_alpha)
+
+  let for_ n ~init ~f =
+    let rec go i acc = if Int.(i = n) then acc else go (i + 1) (f i acc) in
+    go 0 init
+
+  let apply matrix v =
+    let dotv row =
```
[the remark this answers is not in the document]
Comment (imeckler) [reply src/lib/rescue/rescue.ml]: The way I see it `v` is a column and the matrix consists of rows

File: src/lib/rescue/rescue.ml [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+    {mds= f mds; round_constants= f round_constants}
+end
+
+module Make (Inputs : Inputs.S) = struct
+  open Inputs
+
+  let add_block ~state block =
+    Array.iteri block ~f:(fun i bi -> state.(i) <- Field.( + ) state.(i) bi)
+
+  let sponge perm inputs state =
```
[the remark this answers is not in the document]
Comment (imeckler) [reply src/lib/rescue/rescue.ml]: Thanks, I renamed everything to "block" related things

File: src/lib/rescue/rescue.ml (hunk 1 above)
[the remark this answers is not in the document]
Comment (mrmr1993) [reply src/lib/rescue/rescue.ml]: Thanks, that makes perfect sense! I think this might be handy to put in a comment in the code, what do you think?

**Pull request merged**